petab_select/candidate_space.py

```diff
@@ -529,6 +602,628 @@ def iterate():
+                or default_compare(
+                    self.best_model_of_current_run,
+                    local_history[model_id],
+                    criterion,
+                )
+            ):
+                found_new_best = True
+                self.best_model_of_current_run = local_history[model_id]
+
+            if len(self.best_models) < self.most_distant_max_number:
+                self.insert_model_into_best_models(
+                    model_to_insert=local_history[model_id],
+                    criterion=criterion,
+                )
+            elif default_compare(
+                self.best_models[self.most_distant_max_number - 1],
+                local_history[model_id],
+                criterion,
+            ):
+                self.insert_model_into_best_models(
+                    model_to_insert=local_history[model_id],
+                    criterion=criterion,
+                )
+
+        self.best_models = self.best_models[: self.most_distant_max_number]
+
+        # When we switch to LATERAL method, we will do only one iteration with this
+        # method. So if we do it succesfully (i.e. that we found_new_best), we still
+        # want to switch method. This is why we put found_new_best to False, so we go
+        # into the method switching pipeline
+        if (
+            found_new_best
+            and self.method == Method.LATERAL
+            and self.swap_only_once
+        ):
+            self.swap_done_successfully = True
+            found_new_best = False
+        return found_new_best
+
+    def insert_model_into_best_models(
+        self, model_to_insert: Model, criterion: Criterion
+    ) -> None:
+        """Inserts a model into the list of best_models which are sorted
+        w.r.t. the criterion specified."""
+        insert_index = bisect.bisect_left(
+            [model.get_criterion(criterion) for model in self.best_models],
+            model_to_insert.get_criterion(criterion),
+        )
+        self.best_models.insert(insert_index, model_to_insert)
+
+    def consider(self, model: Union[Model, None]) -> bool:
+        """Re-define consider of FAMoS to be the consider method
+        of the inner_candidate_space. Update all the attributes
+        changed in the condsider method."""
+
+        if self.limit.reached():
+            return False
+
+        # Check if model contains necessary critical parameters and, if
+        # the current method is swap, check the swap move is contained
+        # in a swap set.
+
+        if self.check_critical(model) and (
+            not self.method == Method.LATERAL or self.check_swap(model)
+        ):
+            return_value = self.inner_candidate_space.consider(model)
+
+            # update the attributes
+            self.models = self.inner_candidate_space.models
+            self.distances = self.inner_candidate_space.distances
+            self.exclusions = self.inner_candidate_space.exclusions
```

Comment:
This is always some global FAMoS-scope exclusions list right? i.e. it's shared with inner candidate spaces just to update the global exclusions, but is not reset there? i.e. so that it's not possible to visit the same model via the backward method, if it's already been visited by the forward method?

Reply:
If unsure, could append exclusions or do e.g. `list(set(exclusions) | set(new_exclusions))`.

Reply:
Yes, one purpose is so when the method is switched, the new inner_candidate_space gets the exclusions. 

But at this point the main purpose is to maintain self.models of famos to be the same as the ones of the inner_candidate_space. This is because famos's consider method (after checking some things) just calls the consider method of the inner_candidate_space. So the inner_candidate_space considers and accepts (adds into its .models). So we need to accept/exclude all that the inner_candidate_space accepted/excluded. We do this here.